pallets/gear/rpc/src/lib.rs

```diff
@@ -313,34 +444,82 @@ where
     ) -> RpcResult<GasInfo> {
         let at_hash = at.unwrap_or_else(|| self.client.info().best_hash);
 
-        let GasInfo { min_limit, .. } = self.run_with_api_copy(|api| {
-            api.calculate_gas_info(
-                at_hash,
-                source,
-                HandleKind::Reply(
-                    MessageId::from_origin(message_id),
-                    ReplyCode::Success(SuccessReplyReason::Manual),
-                ),
-                payload.to_vec(),
-                value,
-                allow_other_panics,
-                None,
-            )
-        })?;
-        self.run_with_api_copy(|api| {
-            api.calculate_gas_info(
-                at_hash,
-                source,
-                HandleKind::Reply(
-                    MessageId::from_origin(message_id),
-                    ReplyCode::Success(SuccessReplyReason::Manual),
-                ),
-                payload.to_vec(),
-                value,
-                allow_other_panics,
-                Some(min_limit),
-            )
-        })
+        let api_version = self
+            .client
+            .runtime_api()
+            .api_version::<dyn GearRuntimeApi<Block>>(at_hash)
+            .map_err(|e| map_err(e, "Failed to get gear runtime api version"))?
+            .ok_or_else(|| {
+                CallError::Custom(ErrorObject::owned(
+                    8000,
+                    "Gear runtime api wasn't found in the runtime",
+                    None::<String>,
+                ))
+            })?;
+
+        if api_version < 2 {
```

Comment:
We can avoid repeating `self.run_with_api_copy` here.

```rust
let closure = if api_version < 2 {
    |api| {
                #[allow(deprecated)]
                api.calculate_gas_info_before_version_2(...)
    }
else {
    |api| {
                api.calculate_gas_info(...)
    }
};
self.run_with_api_copy(closure);
```

Reply:
moved into trait